src/shims/foreign_items.rs

```diff
@@ -108,6 +126,76 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    /// Lookup the body of a function that has `link_name` as the symbol name.
+    fn lookup_exported_symbol(
+        &mut self,
+        link_name: Symbol,
+    ) -> InterpResult<'tcx, Option<&'mir mir::Body<'tcx>>> {
+        let this = self.eval_context_mut();
+        let tcx = this.tcx.tcx;
+
+        // If the result was cached, just return it.
+        if let Some(instance) = this.machine.exported_symbols_cache.get(&link_name) {
+            return Ok(Some(this.load_mir(instance.def, None)?));
+        }
+
+        // Find it if it was not cached.
+        let mut instance_and_crate: Option<(ty::Instance<'_>, CrateNum)> = None;
+        // `dependency_formats` includes all the transitive informations needed to link a crate,
+        // which is what we need here since we need to dig out `exported_symbols` from all transitive
+        // dependencies.
+        let dependency_formats = tcx.dependency_formats(());
+        let dependency_format = dependency_formats
+            .iter()
+            .find(|(crate_type, _)| *crate_type == CrateType::Executable)
+            .expect("interpreting a non-executable crate");
+        for cnum in
+            iter::once(LOCAL_CRATE).chain(dependency_format.1.iter().enumerate().filter_map(
+                |(num, &linkage)| (linkage != Linkage::NotLinked).then_some(CrateNum::new(num + 1)),
+            ))
+        {
+            // FIXME: Do we need to check `SymbolExportLevel` (the `_` below)?
```

Comment:
Looking at https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/middle/exported_symbols/enum.SymbolExportLevel.html, I don't think we need to check this. We are a Rust crate here, and it looks like everything is exported from a Rust crate.